src/models/source_code_unit.rs

```diff
@@ -339,19 +346,33 @@ impl SourceCodeUnit {
     // Get the tree_sitter's input edit representation
     let (new_source_code, ts_edit) = get_tree_sitter_edit(self.code.clone(), &edit_to_apply);
     // Apply edit to the tree
+    let number_of_errors = self._number_of_errors();
     self.ast.edit(&ts_edit);
     self._replace_file_contents_and_re_parse(&new_source_code, parser, true);
-    if self.ast.root_node().has_error() {
-      let msg = format!(
-        "Produced syntactically incorrect source code {}",
-        self.code()
-      );
-      error!("{}", msg);
-      panic!("{}", msg);
+
+    // Panic if the number of errors increased after the edit
+    if self._number_of_errors() > number_of_errors {
+      self._panic_for_syntax_error();
     }
     ts_edit
   }
 
+  fn _panic_for_syntax_error(&self) {
+    let msg = format!(
+      "Produced syntactically incorrect source code {}",
+      self.code()
+    );
+    error!("{}", msg);
+    panic!("{}", msg);
+  }
+
+  /// Returns the number of errors in the AST
+  fn _number_of_errors(&self) -> usize {
+    traverse(self.root_node().walk(), Order::Post)
+      .filter(|node| node.has_error())
+      .count()
+  }
```

Comment:
One thing that can happen here, which I am not sure how to account for, is a Piranha change making a larger subset of the AST show as an error than before, while retaining the same or smaller error count (e.g. Piranha makes a change that breaks parsing for an entire class which already had at least one syntax error somewhere. Now the entire class shows as an error, and the error count is still 1, but a lot more of the code is unparsable/unanalyzable)